python/ray/autoscaler/_private/autoscaler.py

```diff
@@ -210,66 +218,84 @@ def _update(self):
         last_used = self.load_metrics.last_used_time_by_ip
         horizon = now - (60 * self.config["idle_timeout_minutes"])
 
-        nodes_to_terminate: Dict[NodeID, bool] = []
-        node_type_counts = collections.defaultdict(int)
+        nodes_to_terminate: List[NodeID] = []
+        node_type_counts = defaultdict(int)
         # Sort based on last used to make sure to keep min_workers that
         # were most recently used. Otherwise, _keep_min_workers_of_node_type
         # might keep a node that should be terminated.
         sorted_node_ids = self._sort_based_on_last_used(nodes, last_used)
+
         # Don't terminate nodes needed by request_resources()
-        nodes_allowed_to_terminate: Dict[NodeID, bool] = {}
+        nodes_not_allowed_to_terminate: FrozenSet[NodeID] = {}
         if self.load_metrics.get_resource_requests():
-            nodes_allowed_to_terminate = self._get_nodes_allowed_to_terminate(
-                sorted_node_ids)
+            nodes_not_allowed_to_terminate = \
+                self._get_nodes_needed_for_request_resources(sorted_node_ids)
+
+        def keep_node(node_id: NodeID) -> None:
+            # Update per-type counts and add node_id to nodes_to_keep.
+            tags = self.provider.node_tags(node_id)
+            if TAG_RAY_USER_NODE_TYPE in tags:
+                node_type = tags[TAG_RAY_USER_NODE_TYPE]
+                node_type_counts[node_type] += 1
+
+        def schedule_node_termination(node_id: NodeID, reason: str) -> None:
+            # Log, record an event, and add node_id to nodes_to_terminate.
+            logger.info("StandardAutoscaler: "
+                        "{}: Terminating {} node.".format(node_id, reason))
+            self.event_summarizer.add(
+                "Removing {} nodes of type " + self._get_node_type(node_id) +
+                " ({}).".format(reason),
+                quantity=1,
+                aggregate=operator.add)
+            nodes_to_terminate.append(node_id)
+
+        # Nodes that we could terminate, if needed.
+        nodes_we_could_terminate: List[NodeID] = []
 
         for node_id in sorted_node_ids:
             # Make sure to not kill idle node types if the number of workers
             # of that type is lower/equal to the min_workers of that type
             # or it is needed for request_resources().
-            should_keep_worker_of_node_type, node_type_counts = \
-                self._keep_worker_of_node_type(node_id, node_type_counts)
-            if ((should_keep_worker_of_node_type
-                 or not nodes_allowed_to_terminate.get(node_id, True))
+            should_keep_or_terminate, reason = self._keep_worker_of_node_type(
+                node_id, node_type_counts)
+            if should_keep_or_terminate == KeepOrTerminate.terminate:
+                schedule_node_termination(node_id, reason)
```

Comment:
this generally should not pass mypy typing check because returned type(reason) == Optional[str] but  schedule_node_termination expects str.

Reply:
^That's true.

Though (to clarify) this file is not currently passed through a mypy check.

Reply:
assuming we pass the file to mypy, I could do one of the following:

1. always return and string (empty string instead of None).
2. unwrap the reason into str (check types, raise the exception, assign to a new variable of type str)

1. seems not very explicit and hacky
2. is adding extra boilerplate code (making it less readable)

is there a better way?

Reply:
I think I'd go with the empty string option and document the meaning of empty string reason in the docstring.

Up to you whether you want to make it mypy-compatible -- there aren't any near-term plans to add type-checking here and there are probably loads of other type inconsistencies.

Reply:
On a second thought, I could keep the None and change the signature of schedule_node_termination to
```schedule_node_termination(node_id: NodeID, reason: Optional[str])```

Reply:
that works! -- `None` is a valid input type for string.format